### PR TITLE
[Flink AI Flow] [Plugins]Add flink job plugin

### PR DESCRIPTION
**Pull request by wuchaochen**
## What is the purpose of the change

Add flink job plugin.

### CODE REVIEW

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+
+                if os.path.exists(job_runtime_env.resource_dir):
+                    zip_file_util.make_dir_zipfile(job_runtime_env.resource_dir,
+                                                   os.path.join(job_runtime_env.working_dir, 'resources.zip'))
+                bash_command.extend(['-pyarch',
+                                     os.path.join(job_runtime_env.working_dir, 'resources.zip#resources')])
+                bash_command.extend(['-py', script_path,
+                                     run_graph_file, job_runtime_env.working_dir, flink_env_file])
+            else:
+                raise Exception('Flink supports run_mode loacl or cluster, do not support {}.'
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py:157]:
```suggestion
                raise Exception('Flink supports run_mode local or cluster, do not support {}.'
```

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_config.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+class FlinkJobConfig(JobConfig):
+    """
+    FlinkJobConfig is the configuration of the flink job.
+    job_name:
+        job_type: flink
+        properties:
+            jm: The job manager uri (default None)
+            run_mode: local or cluster (default local)
+            parallelism: The parallelism of the flink job(default 1)
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_config.py:31]: Remove the `parallelism` configuration.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+                bash_command = [python3_location, script_path, run_graph_file, job_runtime_env.working_dir, flink_env_file]
+            elif job_config.run_mode == 'cluster':
+                bash_command = ['flink', 'run']
+                if job_config.jm is not None:
+                    bash_command.extend(['-m', job_config.jm])
+                if job_config.parallelism is not None:
+                    bash_command.extend(['-p', job_config.parallelism])
+
+                if job_config.flink_args is not None:
+                    bash_command.extend(job_config.flink_args)
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py:139]: This should validate the `flink_args` configuration

File: flink-ai-flow/ai_flow_plugins/tests/job_plugins/ut_workflows/workflows/test_flink/test_flink.yaml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+task_1:
+  job_type: flink
+  properties:
+    run_mode: local
+
+task_2:
+  job_type: flink
+  properties:
+    run_mode: cluster
+    parallelism: 1
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/tests/job_plugins/ut_workflows/workflows/test_flink/test_flink.yaml:29]: Remove the `parallelism`.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+                return True
+            else:
+                return False
+        else:
+            for node in sub_graph.nodes.values():
+                processor = node.get_processor()
+                if isinstance(processor, FlinkPythonProcessor):
+                    pass
+                else:
+                    raise Exception("Processor must be instance of FlinkPythonProcessor!")
```
Comment (jiangxin369) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py:81]: raise Exception("A job can only contains one FlinkJavaProcessor or multiple FlinkPythonProcessor.")

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_config.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+from ai_flow.workflow.job_config import JobConfig
+
+
+class FlinkJobConfig(JobConfig):
+    """
+    FlinkJobConfig is the configuration of the flink job.
+    job_name:
+        job_type: flink
+        properties:
+            jm: The job manager uri (default None)
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_config.py:29]: Remove the `jm` configuration.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+                    job_id = fp.read()
+                env = os.environ.copy()
+                env.update(job_config.properties.get('env', {}))
+                # Add PYTHONPATH
+                copy_path = sys.path.copy()
+                copy_path.insert(0, job_runtime_env.python_dep_dir)
+                env['PYTHONPATH'] = ':'.join(copy_path)
+                print('flink cancel {}'.format(job_id))
+                if job_config.jm is None:
+                    bash_command = ['flink', 'cancel', job_id]
```
Comment (jiangxin369) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py:235]: What happened if not assign -jm in cluster mode.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_processor.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+    def process(self, execution_context: ExecutionContext, input_list: List[Table] = None) -> List[Table]:
+        """
+        Process method for user-defined function. User write their logic in this method.
+        """
+        pass
+
+    def setup(self, execution_context: ExecutionContext):
+        """
+        Setup method for user-defined function. It can be used as a preparation stage before the process function.
+        By default, this method does nothing.
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_processor.py:88]: Remove this comment.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_processor.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        """
+        pass
+
+    def close(self, execution_context: ExecutionContext):
+        """
+        Close method for user-defined function. It can be used as a cleanup phase after the process function.
+        By default, this method does nothing.
```
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_processor.py:95]: Remove this comment.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py (hunk 7 above)
Comment (SteNicholas) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py:235]: Stopping flink job should be based on the `flink stop` command.

File: flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+                if 0 == num % 20:
+                    print("Waiting for stdout log file created...")
+            while True:
+                with open(stdout_log, 'r') as f:
+                    lines = f.readlines()
+                    print(lines)
+                if len(lines) >= 1:
+                    line = lines[0]
+                    if line.startswith("Job has been submitted with JobID"):
+                        job_id = line.split(' ')[6][:-1]
```
Comment (jiangxin369) [inline flink-ai-flow/ai_flow_plugins/job_plugins/flink/flink_job_plugin.py:208]: How about using regex to find job_id ?

**Review by SteNicholas — APPROVED** [review]

**Pull request merged**